r/tests/testthat/test-dataset.R

```diff
@@ -1778,3 +1778,60 @@ test_that("Collecting zero columns from a dataset doesn't return entire dataset"
     c(32, 0)
   )
 })
+
+# see https://issues.apache.org/jira/browse/ARROW-12315
+test_that("Max partitions fails with non-integer values and less than required partitions values", {
+  skip_if_not_available("parquet")
+  tmp <- tempfile()
+
+  # this example needs 3 partitions
+
+  # max_partitions = chr => error
+  expect_error(
+    mtcars %>%
+      group_by(cyl) %>%
+      write_dataset(tmp, format = "parquet", max_partitions = "foobar")
+  )
+
+  # max_partitions < 3 => error
+  expect_error(
+    mtcars %>%
+      group_by(cyl) %>%
+      write_dataset(tmp, format = "parquet", max_partitions = -3)
+  )
+
+  # max_partitions < 3 => error
+  expect_error(
+    mtcars %>%
+      group_by(cyl) %>%
+      write_dataset(tmp, format = "parquet", max_partitions = 1)
+  )
```

Comment:
We especially want to make sure that this error is clear + actionable

Reply:
the error in this case is 
```
 Error: Invalid: Fragment would be written into 3 partitions. This exceeds the maximum of 1 
```

I think this is quite clear

Reply:
Yeah, this looks fine too me, let's use the c++ error. If the error is coming directly from c++ (and it's tested there), we don't actually need to test it here as well.